File: src/agenticflow_sdk/exceptions.py
"""Exception hierarchy for the AgenticFlow SDK."""

from __future__ import annotations

from typing import Any


class AgenticFlowError(Exception):
    """Base class for all SDK exceptions."""


class NetworkError(AgenticFlowError):
    """Non-HTTP network-level failure while talking to the API."""

    def __init__(self, message: str, *, cause: Exception | None = None) -> None:
        super().__init__(message)
        self.cause = cause


class RequestTimeoutError(NetworkError):
    """The request exceeded the configured timeout."""


class APIError(AgenticFlowError):
    """Non-successful API response."""

    def __init__(
        self,
        *,
        status_code: int,
        message: str,
        payload: Any | None = None,
        request_id: str | None = None,
    ) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.payload = payload
        self.request_id = request_id


class ValidationError(APIError):
    """Validation or bad request response from the API."""


class AuthenticationError(APIError):
    """Missing or invalid credentials."""


class AuthorizationError(APIError):
    """Authenticated caller does not have access to the resource."""


class NotFoundError(APIError):
    """Target resource could not be found."""


class ConflictError(APIError):
    """Request conflicts with current resource state."""


class RateLimitError(APIError):
    """API rate limits were hit."""


class ServerError(APIError):
    """A 5xx response was returned by the API."""

File: packages/sdk/src/exceptions.ts
/**
 * Exception hierarchy for the AgenticFlow SDK.
 */

export class AgenticFlowError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AgenticFlowError";
  }
}

export class NetworkError extends AgenticFlowError {
  cause?: Error;

  constructor(message: string, options?: { cause?: Error }) {
    super(message);
    this.name = "NetworkError";
    this.cause = options?.cause;
  }
}

export class RequestTimeoutError extends NetworkError {
  constructor(message: string = "Request timed out.", options?: { cause?: Error }) {
    super(message, options);
    this.name = "RequestTimeoutError";
  }
}

export class APIError extends AgenticFlowError {
  statusCode: number;
  payload: unknown;
  requestId: string | null;

  constructor(options: {
    statusCode: number;
    message: string;
    payload?: unknown;
    requestId?: string | null;
  }) {
    super(options.message);
    this.name = "APIError";
    this.statusCode = options.statusCode;
    this.payload = options.payload ?? null;
    this.requestId = options.requestId ?? null;
  }
}

export class ValidationError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "ValidationError";
  }
}

export class AuthenticationError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "AuthenticationError";
  }
}

export class AuthorizationError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "AuthorizationError";
  }
}

export class NotFoundError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "NotFoundError";
  }
}

export class ConflictError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "ConflictError";
  }
}

export class RateLimitError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "RateLimitError";
  }
}

export class ServerError extends APIError {
  constructor(options: ConstructorParameters<typeof APIError>[0]) {
    super(options);
    this.name = "ServerError";
  }
}
